File: problems/scaler/Prefix_sum/Special_index.js
/**
 * Given an array, find the count of array indices such that removing an element from these indices
 * makes the sum of even-indexed and odd-indexed array elements equal.
 *
 * @param {number[]} A - Input array
 * @return {number} - Count of special indices
 */
function countSpecialIndices(A) {
  const n = A.length;
  const evenPrefix = new Array(n).fill(0);
  const oddPrefix = new Array(n).fill(0);

  evenPrefix[0] = A[0];
  oddPrefix[0] = 0;

  // Calculate prefix sums
  for (let i = 1; i < n; i++) {
    if (i % 2 === 0) {
      evenPrefix[i] = evenPrefix[i - 1] + A[i];
      oddPrefix[i] = oddPrefix[i - 1];
    } else {
      evenPrefix[i] = evenPrefix[i - 1];
      oddPrefix[i] = oddPrefix[i - 1] + A[i];
    }
  }

  let count = 0;
  // Check each index
  for (let i = 0; i < n; i++) {
    let evenSum = 0;
    let oddSum = 0;

    // Calculate even sum after removing element at index i
    if (i === 0) {
      evenSum = oddPrefix[n - 1];
      oddSum = evenPrefix[n - 1] - A[0];
    } else {
      evenSum = evenPrefix[i - 1] + (oddPrefix[n - 1] - oddPrefix[i]);
      oddSum = oddPrefix[i - 1] + (evenPrefix[n - 1] - evenPrefix[i]);
    }

    if (evenSum === oddSum) {
      count++;
    }
  }

  return count;
}

// Test cases
console.log(countSpecialIndices([2, 1, 6, 4])); // Output: 1
console.log(countSpecialIndices([1, 1, 1])); // Output: 3
console.log(countSpecialIndices([1, 2, 3, 7, 1, 2, 3])); // Output: 1

File: src/main/java/com/dsa/problems/scaler/prefix_sum/Special_Index.java
/**
 * Given an array, find the count of array indices such that removing an element from these indices
 * makes the sum of even-indexed and odd-indexed array elements equal.
 *
 */
package com.dsa.problems.scaler.prefix_sum;

import java.util.ArrayList;
import java.util.Scanner;

public class Special_Index {
  static public ArrayList<Integer> getPrefixSum(ArrayList<Integer> A, String type) {
    ArrayList<Integer> prefixSum = new ArrayList<>();

    int n = A.size();
    for (int i = 0; i < n; i++) {
      if (type.equals("even")) {
        if (i % 2 == 0) {
          if (i == 0) {
            prefixSum.add(A.get(i));
          } else {
            prefixSum.add(prefixSum.get(i - 1) + A.get(i));
          }
        } else {
          prefixSum.add(prefixSum.get(i - 1));
        }
      } else if (type.equals("odd")) {
        if (i % 2 == 1) {
          prefixSum.add(prefixSum.get(i - 1) + A.get(i));
        } else {
          if (i == 0) {
            prefixSum.add(0);
          } else {
            prefixSum.add(prefixSum.get(i - 1));
          }
        }
      }
    }

    return prefixSum;
  }

  public static int countSpecialIndices(ArrayList<Integer> A) {
    int n = A.size();
    ArrayList<Integer> evenPrefixSum = getPrefixSum(A, "even");
    ArrayList<Integer> oddPrefixSum = getPrefixSum(A, "odd");

    int count = 0;
    for (int i = 0; i < n; i++) {
      if (i == 0) {
        if (evenPrefixSum.get(n - 1) - A.get(0) == oddPrefixSum.get(n - 1)) {
          count++;
        }
      } else {
        int newEvenSum = evenPrefixSum.get(i - 1) + oddPrefixSum.get(n - 1) - oddPrefixSum.get(i);
        int newOddSum = oddPrefixSum.get(i - 1) + evenPrefixSum.get(n - 1) - evenPrefixSum.get(i);
        if (newOddSum == newEvenSum) {
          count++;
        }
      }
    }

    return count;
  }

  public static void main(String[] args) {
    try (Scanner sc = new Scanner(System.in)) {
      int n = sc.nextInt();
      ArrayList<Integer> A = new ArrayList<>();
      for (int i = 0; i < n; i++) {
        A.add(sc.nextInt());
      }
      System.out.println(countSpecialIndices(A));
    }
  }
}
